core/arch/x86/decode_fast.c

```diff
@@ -1352,14 +1352,21 @@ decode_cti(dcontext_t *dcontext, byte *pc, instr_t *instr)
     }
     /* popf */
     if (byte0 == 0x9d) {
+        reg_id_t stack_sized_reg = REG_XSP;
+#ifdef X64
+        if (dr_get_isa_mode(dcontext) == DR_ISA_IA32) {
+            stack_sized_reg = REG_ESP;
+        }
+#endif
         instr_set_opcode(instr, OP_popf);
         instr_set_raw_bits(instr, start_pc, sz);
         instr_set_num_opnds(dcontext, instr, 1, 2);
-        instr_set_src(instr, 0, opnd_create_reg(REG_XSP));
+        instr_set_src(instr, 0, opnd_create_reg(stack_sized_reg));
         instr_set_src(instr, 1, opnd_create_base_disp
-                      (REG_XSP, REG_NULL, 0, 0,
-                       resolve_variable_size_dc(dcontext, 0, OPSZ_ret, false)));
-        instr_set_dst(instr, 0, opnd_create_reg(REG_XSP));
+                      (stack_sized_reg, REG_NULL, 0, 0,
+                       resolve_variable_size_dc(dcontext, 0,
```

Comment:
OK, so it goes to the slow path.  It doesn't have to for this case but it's not worth reordering.  It still seems safer to pass prefixes here rather than 0 to future-proof.  It would also be nice to have a test w/ a size prefix but there is a long list of "nice to have" things in the tests so this seems sufficient.